R/facet-.r

```diff
@@ -203,38 +264,152 @@ df.grid <- function(a, b) {
   ))
 }
 
+# A facets spec is a list of facets. A grid facetting needs two facets
+# while a wrap facetting flattens all dimensions and thus accepts any
+# number of facets.
+#
+# A facets is a list of grouping variables. They are typically
+# supplied as variable names but can be expressions.
+#
+# as_facets() is complex due to historical baggage but its main
+# purpose is to create a facets spec from a formula: a + b ~ c + d
+# creates a facets list with two components, each of which bundles two
+# facetting variables.
+
+as_facets_list <- function(x) {
+  if (inherits(x, "mapping")) {
+    stop("Please use `vars()` to supply facet variables")
+  }
+  if (inherits(x, "quosures")) {
+    x <- rlang::quos_auto_name(x)
+    return(list(x))
+  }
+
+  # This needs to happen early because we might get a formula.
+  # facet_grid() directly converted strings to a formula while
+  # facet_wrap() called as.quoted(). Hence this is a little more
+  # complicated for backward compatibility.
+  if (rlang::is_string(x)) {
+    x <- plyr::as.quoted(x)
+    if (rlang::is_formula(x[[1]])) {
+      x <- x[[1]]
+    }
+  }
+
+  # At this level formulas are coerced to lists of lists for backward
+  # compatibility with facet_grid(). The LHS and RHS are treated as
+  # distinct facet dimensions and `+` defines multiple facet variables
+  # inside each dimension.
+  if (rlang::is_formula(x)) {
+    return(f_as_facets_list(x))
+  }
+
+  # For backward-compatibility with facet_wrap()
+  if (!rlang::is_bare_list(x)) {
+    x <- plyr::as.quoted(x)
+    attributes(x) <- NULL
+  }
+
+  # If we have a list there are two possibilities. We may already have
+  # a proper facet spec structure. Otherwise we coerce each element
+  # with plyr::as.quoted() for backward compatibility with facet_grid().
+  if (is.list(x)) {
+    x <- lapply(x, as_facets)
+  }
+
+  if (sum(vapply(x, length, integer(1))) == 0L) {
+    stop("Must specify at least one variable to facet by", call. = FALSE)
+  }
+
+  x
+}
+
+f_as_facets_list <- function(f) {
+  lhs <- function(x) if (length(x) == 2) NULL else x[-3]
+  rhs <- function(x) if (length(x) == 2) x else x[-2]
+
+  rows <- f_as_facets(lhs(f))
+  cols <- f_as_facets(rhs(f))
+
+  if (length(rows) + length(cols) == 0) {
+    stop("Must specify at least one variable to facet by", call. = FALSE)
+  }
+
+  if (length(rows)) {
+    list(rows, cols)
+  } else {
+    list(cols)
+  }
+}
+
+as_facets <- function(x) {
+  if (is_facets(x)) {
+    return(x)
+  }
+
+  if (rlang::is_formula(x)) {
+    # Use different formula method because plyr's does not handle the
+    # environment correctly.
+    f_as_facets(x)
+  } else {
+    vars <- plyr::as.quoted(x)
+    rlang::as_quosures(vars, globalenv(), named = TRUE)
+  }
+}
+f_as_facets <- function(f) {
+  if (is.null(f)) {
+    return(rlang::as_quosures(list()))
+  }
+
+  env <- rlang::f_env(f) %||% globalenv()
+
+  # as.quoted() handles `+` specifications
+  vars <- plyr::as.quoted(f)
```

Comment:
How much work would it be to eliminate `as.quoted()` altogether?

Reply:
If we handled the formula decomposition (turning an addition of expressions to a list of quosures) that would take care of the majority of the cases, but `as.quoted()` is a S3 generic so for backward compat I thought it was best to keep it. Do you want to remove it?

Reply:
Remove it - it's esoteric.